### ISSUE TITLE
 can't delete certain voicemails

### ISSUE THREAD

**Issue by alanb2718** [bmlt-enabled/yap#576]
There are some empty voicemails for Seattle Area yap, running on the WSZF server with yap 4.0.1.  If I press the "delete" button, they don't get deleted; they just sit there.  However, other voicemails _can_ be deleted.  Then I turned on the javascript console to see if that would turn up anything.  This shows an error when I first start up the display.  However, I don't see an error when I press delete, so maybe this error is unrelated.

I then tried this on our test server, and _was_ able to delete voicemails, including empty ones.  But the javascript console shows that same error when I first start up the display.  Here's the error I keep seeing on the Javascript console:

```
Uncaught SyntaxError: Unexpected token u in JSON at position 0
    at JSON.parse (<anonymous>)
    at Q (yap-reports.min.js:1)
    at Q (yap-reports.min.js:1)
    at Q (yap-reports.min.js:1)
    at make_xlsx_lib (yap-reports.min.js:1)
    at yap-reports.min.js:1

```

Here's what larval says:
```

[2021-12-07 06:23:28] laravel.EMERGENCY: Unable to create configured logger. Using emergency logger. {"exception":"[object] (InvalidArgumentException(code: 0): Log [] is not defined. at /home/customer/www/bmlt.wszf.org/public_html/yap/vendor/laravel/framework/src/Illuminate/Log/LogManager.php:192)
[stacktrace]
#0 /home/customer/www/bmlt.wszf.org/public_html/yap/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(118): Illuminate\\Log\\LogManager->resolve(NULL)
#1 /home/customer/www/bmlt.wszf.org/public_html/yap/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(98): Illuminate\\Log\\LogManager->get(NULL)
#2 /home/customer/www/bmlt.wszf.org/public_html/yap/vendor/laravel/framework/src/Illuminate/Log/LogManager.php(555): Illuminate\\Log\\LogManager->driver()
#3 /home/customer/www/bmlt.wszf.org/public_html/yap/vendor/laravel/framework/src/Illuminate/Foundation/Exceptions/Handler.php(252): Illuminate\\Log\\LogManager->error('Class 'SearchTy...', Array)
#4 /home/customer/www/bmlt.wszf.org/public_html/yap/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(414): Illuminate\\Foundation\\Exceptions\\Handler->report(Object(Error))
#5 /home/customer/www/bmlt.wszf.org/public_html/yap/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(112): Illuminate\\Foundation\\Http\\Kernel->reportException(Object(Error))
#6 /home/customer/www/bmlt.wszf.org/public_html/yap/index.php(52): Illuminate\\Foundation\\Http\\Kernel->handle(Object(Illuminate\\Http\\Request))
#7 {main}
"} 
[2021-12-07 06:23:28] laravel.ERROR: Class 'SearchType' not found {"exception":"[object] (Error(code: 0): Class 'SearchType' not found at /home/customer/www/bmlt.wszf.org/public_html/yap/config.php:15)
[stacktrace]
#0 /home/customer/www/bmlt.wszf.org/public_html/yap/config/database.php(3): require_once()
#1 /home/customer/www/bmlt.wszf.org/public_html/yap/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/LoadConfiguration.php(72): require('/home/customer/...')
#2 /home/customer/www/bmlt.wszf.org/public_html/yap/vendor/laravel/framework/src/Illuminate/Foundation/Bootstrap/LoadConfiguration.php(39): Illuminate\\Foundation\\Bootstrap\\LoadConfiguration->loadConfigurationFiles(Object(Illuminate\\Foundation\\Application), Object(Illuminate\\Config\\Repository))
#3 /home/customer/www/bmlt.wszf.org/public_html/yap/vendor/laravel/framework/src/Illuminate/Foundation/Application.php(237): Illuminate\\Foundation\\Bootstrap\\LoadConfiguration->bootstrap(Object(Illuminate\\Foundation\\Application))
#4 /home/customer/www/bmlt.wszf.org/public_html/yap/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(152): Illuminate\\Foundation\\Application->bootstrapWith(Array)
#5 /home/customer/www/bmlt.wszf.org/public_html/yap/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(136): Illuminate\\Foundation\\Http\\Kernel->bootstrap()
#6 /home/customer/www/bmlt.wszf.org/public_html/yap/vendor/laravel/framework/src/Illuminate/Foundation/Http/Kernel.php(110): Illuminate\\Foundation\\Http\\Kernel->sendRequestThroughRouter(Object(Illuminate\\Http\\Request))
#7 /home/customer/www/bmlt.wszf.org/public_html/yap/index.php(52): Illuminate\\Foundation\\Http\\Kernel->handle(Object(Illuminate\\Http\\Request))
#8 {main}
"} 

```

**Comment by alanb2718** [comment]
We have been using the delete functionality to coordinate volunteers who return voicemails (someone deletes the voicemail to indicate it's been answered).  If it can be done safely, I'd like to edit our database manually to remove the empty voicemails for now.  It looks like normally a delete operation is reflected in the database by a record in the `event_status` table.  There isn't a record created when I try unsuccessfully to delete a voicemail.  So .... would it be safe to just delete the records from `records_events` for the voicemails that can't be deleted?  I've got a backup of the database in case examples of this bug are needed later for debugging.

**Comment by dgershman** [comment]
I believe I understand what the issue is.  Can you give more information about the user you are using?  Is this a BMLT user or a Yap user?  Do they have to be a special admin user?


I will not make recommendations to modify the database like this.

**Comment by alanb2718** [comment]
The user is just the normal SeattleArea BMLT user.

**Comment by dgershman** [comment]
What permissions does the user have?  I'm assuming they see the service body in service bodies, and this is also the same place where they click "Voicemail".  Is there a 403 response code from the request when you click delete in your debugging tools?

**Comment by alanb2718** [comment]
As far as BMLT is concerned, the SeattleArea user is a service body administrator, owned by Washington/Northern Idaho Region. On the yap interface, this user sees the Seattle Area service body.  This is where I'm clicking "Voicemail".  There isn't a 403 response code at least that I can see.  In the javascript console there is the uncaught syntax error reported earlier.  There is also a message "Depreciation: Synchronous HTTP something request ..." that flits by so fast, and is then overwritten with the syntax error exception, that I can't read any more of it.

**Comment by alanb2718** [comment]
I tried making a separate yap user, and it still can't delete those voicemails.

**Comment by dgershman** [comment]
When you said you made a separate yap user what do you mean by that?  Did you make an admin user or a regular user?

**Comment by dgershman** [comment]
When I said 403 response code, I was referring to the typical Network tab on your developer tools.  You can click "Persist..." in order keep the logs from refreshing on a page load (which is what is happening).

**Comment by dgershman** [comment]
Ok I think I know what the issue is.

**Comment by alanb2718** [comment]
By "separate yap user" I mean the user created by following the directions in the yap documentation.  This is an admin user. 

**Users**
User authentication has primarily been sourced by a BMLT. You can create users outside the BMLT and use Yap’s internal user authentication. To create an admin user run the following MySQL script. Be sure to set a strong password and fill it in the variable before running this on your MySQL instance.
```
SET @realname = '';
SET @username = '';
SET @password = '';
INSERT INTO users (name, username, password, permissions, is_admin) VALUES (@realname, @username, SHA2(@password, 256), 0, 1);
```

**Comment by dgershman** [comment]
Thanks.  I believe I know the root cause now.

**Comment by dgershman** [comment]
fix for this will be 4.0.2, hopefully will release it today

**Comment by alanb2718** [comment]
Sorry, this is still not working in 4.0.2

**Comment by dgershman** [comment]
Please define not working.

**Comment by alanb2718** [comment]
Exactly the same behavior as before.  I try to delete the voicemail and nothing happens.  Same javascript error as in the original report.

**Comment by dgershman** [comment]
Ok.  Thanks.

**Comment by dgershman** [comment]
Can you let me know the HTTP response code on the request for `/v1/events/status`?  That request is made when you click the Delete button in a voicemail.  

You can see this by setting in Chrome Developer Tools, then click the Network tab and then click Preserve Log (as seen in the screen shot).  If it's a 5xx error then send along the accompanying lararvel.log entries.  It could also be a 403 error.

![image](https://user-images.githubusercontent.com/1850811/146850256-2a90fd96-3572-42f7-81f1-1dccce9b763d.png)

**Comment by alanb2718** [comment]
Here is the results of checking Preserve Log -- there is a 500 error.  Also attached is the relevant part of laravel.log.

![yap-error](https://user-images.githubusercontent.com/918773/147005409-a76494e6-9f9e-4719-adfb-61b94c6f5a1a.jpg)
[laravel.log](https://github.com/bmlt-enabled/yap/files/7758688/laravel.log)

**Comment by dgershman** [comment]
Do you have a value set in your config file for `$digit_map_search_type`/

**Comment by alanb2718** [comment]
Yes, it's set to 
`static $digit_map_search_type = ['1' => SearchType::VOLUNTEERS, '2' => SearchType::MEETINGS, '3' => SearchType::JFT, '8' => SearchType::VOICEMAIL_PLAYBACK];`

This was the setting you recommended to work around the security issue for dialing 9 (#527).  I just tried removing it, though, and still see exactly the same behavior.

**Comment by dgershman** [comment]
I was not suggesting anything, I was only asking a question.  

After you disable it do you see a different error message in your log file?

**Comment by alanb2718** [comment]
When I remove the line that sets `$digit_map_search_type` in `config.php`, I still see the same 500 error for the HTTP response code.  The laravel log is different; I'm not sure if it's a meaningful difference though.  The latest log is attached.  Thanks for all your work in trying to track this down!!
[newlaravel.log](https://github.com/bmlt-enabled/yap/files/7764977/newlaravel.log)

**Comment by dgershman** [comment]
@alanb2718 This build should address your issues https://s3.amazonaws.com/archives.bmlt.app/yap/yap-85-c5c32c7c149cfc3e2928cfc13e82b95bd8628c97.zip.  Can you test it ?

**Comment by alanb2718** [comment]
It works - thanks!

**Comment by dgershman** [comment]
Available in 4.0.3
